telemetry/class-telemetry-event.php

```diff
@@ -16,12 +16,59 @@
  * Base class for all telemetry event implementations.
  */
 abstract class Telemetry_Event implements JsonSerializable {
+	/**
+	 * Variable containing the event's data or a WP_Error if an error was
+	 * encountered during the event's creation or validation.
+	 *
+	 * @var object|WP_Error
+	 */
+	protected object $data;
+
+	/**
+	 * Generate the event data.
+	 *
+	 * @return object|WP_Error A serializable object if the event is valid, otherwise a WP_Error
+	 */
+	abstract protected function generate(): object;
+
+	/**
+	 * Wraps generate() and stores the result to prevent multiple calls.
+	 *
+	 * @return object|WP_Error A serializable object if the event is valid, otherwise a WP_Error
+	 */
+	public function get_data(): object {
+		if ( ! isset( $this->data ) ) {
+			$this->data = $this->generate();
+		}
+
+		return $this->data;
+	}
 
 	/**
 	 * Returns whether the event can be recorded.
 	 *
 	 * @return bool|WP_Error True if the event is recordable.
-	 *                        WP_Error is any error occurred.
 	 */
-	abstract public function is_recordable();
+	public function is_recordable(): bool|WP_Error {
+		$data = $this->get_data();
+
+		if ( is_wp_error( $data ) ) {
+			return $data;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Returns the event's data for JSON representation.
+	 */
+	public function jsonSerialize(): mixed {
```

Comment:
nitpick but violating WP naming conventions for functions and methods (snake_case over camelCase)

Reply:
Good call out. However, this method is implementing the `JsonSerializable` interface from PHP core so that we can customize the behavior of `json_encode`. Also, this is relocated code from `Tracks_Event`.

https://www.php.net/manual/en/class.jsonserializable.php